### ISSUE TITLE
Spike: Identify Pattern For API Security

### ISSUE THREAD

**Issue by ArogeG** [bcgov/nr-forests-access-management#277]
Before implementing the FAM API Security, explore options for security so that appropriate patterns can be implemented for the security into FAM API. 

- Consider using FAST API patterns. 
- Another Option is the python library. 
- Supports local API testing

https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/

**Comment by ConradBoydElliottGustafson** [comment]
Verified that this tutorial is wrong. It's setting up FastAPI as an OAuth server. What we want is to consume OIDC tokens as an OIDC client

**Comment by ConradBoydElliottGustafson** [comment]
This library looks promising, but it seems to only deal with ID tokens. We care about access tokens. https://pypi.org/project/fastapi-oidc/

**Comment by franTarkenton** [comment]
Not sure if this will help, but its what I had bookmarked from a while ago when I was trying to figure this out... https://dev.to/livioribeiro/fastapi-with-oidc-discovery-4507

**Comment by franTarkenton** [comment]
... and a few more from the last time I tried to wrap my head around this, not sure if they will help but sending along just in case:
* https://github.com/tiangolo/full-stack-fastapi-postgresql
* https://github.com/andreaskring/fastapi-keycloak-angular

I think this shows a basic pattern for how it might work... 
https://github.com/andreaskring/fastapi-keycloak-angular/blob/master/fastapi/main.py

**Comment by ConradBoydElliottGustafson** [comment]
OK, I think I have all the necessary information now. The FastAPI security library has a really nice feature that adds a check for the bearer token and passes the token to the router method. Setting this up integrates with the /docs endpoint and the OpenAPI specification that is generated by FastAPI. It's slick. I tested it out and it works.

What this feature does NOT do is validate the token. We should be checking that the token is correctly signed, that it is meant for the correct audience (FAM), and that it has not expired. We will have to code these checks, but I know of python libraries that I can steal the code from.

I hunted around for any other options, but they were all much weaker than this and would require just as much coding (probably more).
